File: web-frontend/src/types/notification.ts
/**
 * Notification types for BATbern platform
 * Based on Story BAT-7 (Notification System) and EventBridge integration
 */

export type NotificationType =
  // Event notifications
  | 'EVENT_PUBLISHED'
  | 'EVENT_STATUS_CHANGED'
  // Speaker workflow notifications
  | 'SPEAKER_INVITED'
  | 'SPEAKER_ACCEPTED'
  | 'SPEAKER_DECLINED'
  | 'SPEAKER_CONFIRMED'
  | 'CONTENT_SUBMITTED'
  | 'QUALITY_REVIEW_PENDING'
  | 'QUALITY_REVIEW_APPROVED'
  | 'QUALITY_REVIEW_REQUIRES_CHANGES'
  | 'SLOT_ASSIGNED'
  // Task notifications
  | 'TASK_ASSIGNED'
  | 'TASK_DEADLINE_WARNING'
  | 'DEADLINE_WARNING'
  // Other notifications
  | 'OVERFLOW_DETECTED'
  | 'VOTING_REQUIRED'
  | 'PARTNER_ANALYSIS_READY'
  | 'NEW_MESSAGE'
  | 'SYSTEM_ANNOUNCEMENT';

export type NotificationPriority = 'LOW' | 'NORMAL' | 'HIGH' | 'URGENT';

export type NotificationStatus = 'PENDING' | 'SENT' | 'FAILED' | 'READ' | 'UNREAD';

export type NotificationChannel = 'EMAIL' | 'SMS' | 'WEBHOOK' | 'IN_APP';

export interface Notification {
  id: string;
  recipientUsername: string;
  eventCode?: string; // Nullable for non-event notifications
  notificationType: NotificationType;
  channel: NotificationChannel;
  priority: NotificationPriority;
  subject: string;
  body: string;
  status: NotificationStatus;
  sentAt?: string; // ISO 8601 format
  readAt?: string; // ISO 8601 format
  failedAt?: string; // ISO 8601 format
  failureReason?: string;
  metadata?: Record<string, unknown>;
  createdAt: string; // ISO 8601 format
  updatedAt: string; // ISO 8601 format
  // UI helper properties (aliases for backward compatibility)
  isRead?: boolean; // Computed from status === 'READ'
  actionUrl?: string; // Computed from eventCode/metadata
  title?: string; // Alias for subject (for UI components)
  message?: string; // Alias for body (for UI components)
  timestamp?: string; // Alias for createdAt (for UI components)
}

export interface NotificationsResponse {
  data: Notification[];
  pagination: {
    page: number;
    limit: number;
    totalItems: number;
    totalPages: number;
  };
  unreadCount?: number; // Total unread notifications (for in-app notifications)
}

export interface NotificationCountResponse {
  count: number;
}

export interface MarkAsReadRequest {
  notificationIds?: string[];
}

export interface MarkAsReadResponse {
  success: boolean;
  markedCount: number;
  updatedAt: string;
}

export interface DeleteNotificationResponse {
  success: boolean;
}

export interface NotificationFilters {
  username: string;
  status?: 'PENDING' | 'SENT' | 'READ' | 'FAILED' | 'UNREAD';
  eventCode?: string;
  notificationType?: NotificationType;
}

File: services/event-management-service/src/main/java/ch/batbern/events/notification/Notification.java
package ch.batbern.events.notification;

import io.hypersistence.utils.hibernate.type.json.JsonBinaryType;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.Type;

import java.time.Instant;
import java.util.Map;
import java.util.UUID;

/**
 * Notification entity
 * Story BAT-7: Notifications API Consolidation
 *
 * Follows Task System pattern (Story 5.5):
 * - Single table in Event Management Service
 * - ADR-003 compliant (meaningful IDs: recipient_username, event_code)
 * - Hybrid storage (email creates rows, in-app queries dynamic)
 */
@Entity
@Table(name = "notifications")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Notification {

    @Id
    @GeneratedValue
    private UUID id;

    // ADR-003: Meaningful IDs (NOT foreign keys)
    @Column(name = "recipient_username", nullable = false, length = 100)
    private String recipientUsername;

    @Column(name = "event_code", length = 50)
    private String eventCode;  // Nullable for non-event notifications

    // Notification details
    @Column(name = "notification_type", nullable = false, length = 50)
    private String notificationType;

    @Column(name = "channel", nullable = false, length = 20)
    private String channel;  // EMAIL, SMS

    @Column(name = "priority", length = 20)
    private String priority;  // LOW, NORMAL, HIGH, URGENT

    @Column(name = "subject", nullable = false)
    private String subject;

    @Column(name = "body", nullable = false, columnDefinition = "TEXT")
    private String body;

    // Delivery tracking
    @Column(name = "status", nullable = false, length = 20)
    private String status;  // PENDING, SENT, FAILED, READ

    @Column(name = "sent_at")
    private Instant sentAt;

    @Column(name = "read_at")
    private Instant readAt;

    @Column(name = "failed_at")
    private Instant failedAt;

    @Column(name = "failure_reason", columnDefinition = "TEXT")
    private String failureReason;

    // Metadata (flexible JSONB storage)
    @Type(JsonBinaryType.class)
    @Column(name = "metadata", columnDefinition = "jsonb")
    private Map<String, Object> metadata;

    // Timestamps
    @Column(name = "created_at", nullable = false)
    private Instant createdAt;

    @Column(name = "updated_at", nullable = false)
    private Instant updatedAt;

    @PrePersist
    protected void onCreate() {
        createdAt = Instant.now();
        updatedAt = Instant.now();
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = Instant.now();
    }
}
